code/src/sixsq/nuvla/server/resources/nuvlabox/utils.clj

```diff
@@ -106,3 +107,82 @@
 (defn get-execution-mode
   [nuvlabox]
   (if (has-pull-support? nuvlabox) "pull" "push"))
+
+
+(defn get-playbooks
+  ([nuvlabox-id] (get-playbooks nuvlabox-id "MANAGEMENT"))
+  ([nuvlabox-id type]
+   (if-not (or (empty? nuvlabox-id) (empty? type))
+     (->> {:params      {:resource-name "nuvlabox-playbook"}
+           :cimi-params {:filter (parser/parse-cimi-filter
+                                   (format "enabled='true' and parent='%s' and type='%s'" nuvlabox-id type))
+                         :select ["run", "id"]
+                         :last 1000}
+           :nuvla/authn auth/internal-identity}
+       crud/query
+       :body
+       :resources)
+     [])))
+
+
+(defn get-nuvlabox-playbook-filename
+  [id]
+  (str "/tmp/nuvlabox-playbook-" (last (str/split id #"/"))))
+
+
+(defn get-nuvlabox-playbook-output-filename
+  [id]
+  (str (get-nuvlabox-playbook-filename (last (str/split id #"/"))) ".output"))
+
+
+(defn wrap-playbook-run
+  [playbook]
+  (let [id  (last (str/split (:id playbook) #"/"))
+        nuvlabox-playbook-file (get-nuvlabox-playbook-filename id)
+        nuvlabox-playbook-out  (get-nuvlabox-playbook-output-filename id)]
+  (str
+    "cat > " nuvlabox-playbook-file " <<'EOF'\n"
+    (:run playbook)
+    "\nEOF\n"
+    "echo '' > " nuvlabox-playbook-out
+    "\nsh " nuvlabox-playbook-file
+    " 2>&1 | while IFS= read -r line; do printf '[%s] %s\\n' \"$(date '+%Y-%m-%d %H:%M:%S')\" \"$line\" >> "
+    nuvlabox-playbook-out "; done || true")))
+
+
+(def ^:const nuvla-login-script (str "curl -X POST ${NUVLA_ENDPOINT:-https://nuvla.io}/api/session "
+                                  "-H content-type:application/json "
+                                  "-c /tmp/nuvla-cookie "
+                                  "-d \"{\\\"template\\\": {\\\"href\\\": \\\"session-template/api-key\\\",\\\"key\\\": \\\"$NUVLABOX_API_KEY\\\", \\\"secret\\\": \\\"$NUVLABOX_API_SECRET\\\"}}\""))
+
+
+(defn wrap-and-pipe-playbooks
+  [playbooks]
+  (if (empty? playbooks)
+    ""
+    (let [wrapped-runs      (for [playbook playbooks] (wrap-playbook-run playbook))
```

Comment:
why not use map instead of for? When you are looping a single sequence and the result is another sequence of the same size, usually we use map. For can treat multiple sequences combining them.

Reply:
changed